tracer/src/Datadog.Trace/PDBs/SourceLinkInformationExtractor.cs

```diff
@@ -0,0 +1,90 @@
+﻿// <copyright file="SourceLinkInformationExtractor.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Reflection;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Pdb.SourceLink;
+using Datadog.Trace.Vendors.Newtonsoft.Json.Linq;
+
+#nullable enable
+
+namespace Datadog.Trace.Pdb;
+
+internal static class SourceLinkInformationExtractor
+{
+    public static bool TryGetSourceLinkInfo(Assembly assembly, [NotNullWhen(true)] out string? commitSha, [NotNullWhen(true)] out string? repositoryUrl)
+    {
+        // Extracting the SourceLink information from the assembly attributes will only work if:
+        // 1. The assembly was built using the .NET Core SDK 2.1.300 or newer or MSBuild 15.7 or newer.
+        // 2. The assembly was built using an SDK-Style project file (i.e. <Project Sdk="Microsoft.NET.Sdk">).
+        // If these conditions weren't met, the attributes won't be there, so we'll need to extract the information from the PDB file.
+
+        return ExtractFromAssemblyAttributes(assembly, out commitSha, out repositoryUrl) ||
+               ExtractFromPdb(assembly, out commitSha, out repositoryUrl);
+    }
+
+    private static bool ExtractFromPdb(Assembly assembly, out string? commitSha, out string? repositoryUrl)
+    {
+        commitSha = null;
+        repositoryUrl = null;
+        var sourceLinkJsonDocument = DatadogPdbReader.CreatePdbReader(assembly)?.GetSourceLinkJsonDocument();
+        if (sourceLinkJsonDocument == null)
+        {
+            return false;
+        }
+
+        // Grab the SourceLink mapping URL. For example,
+        // From:
+        //       {"documents":{"C:\\dev\\dd-trace-dotnet\\*":"https://raw.githubusercontent.com/DataDog/dd-trace-dotnet/dd35903c688a74b62d1c6a9e4f41371c65704db8/*"}}
+        // Extract:
+        //       https://raw.githubusercontent.com/DataDog/dd-trace-dotnet/dd35903c688a74b62d1c6a9e4f41371c65704db8/*
+        var sourceLinkMappedUrl = JObject.Parse(sourceLinkJsonDocument).SelectTokens("$.documents.*").FirstOrDefault()?.ToString();
```

Comment:
This could throw if it's malformed JSON - should we put in a `try-catch`? _Maybe_ also go with simple foreach rather than linq (just because that's what generally advocate, not passionate about it here, little impact I'm sure!)

Reply:
Added a `try/catch` in 233cf69448d27b82b4c1721ca90bb74fab58032f.  
This code should only be executed _once_ in the entire lifetime of the application, so I kept the LINQ usage as the perf impact should be quite negligible.